utils/auction/clear-auction.js

```diff
@@ -0,0 +1,189 @@
+const fs = require('fs');
+const path = require('path');
+
+const USDC_SCALE = 1_000_000n;                // 1 USDC = 1e6
+
+/**
+ * Runs a uniform-price auction using BigInt with:
+ *   - USDC in 6 decimals (1 USDC = 1e6)
+ *   - Tokens in 18 decimals (1 TOKEN = 1e18)
+ *
+ * @param {Object[]} bids
+ * @param {string}   bids[].address      - Ethereum address.
+ * @param {bigint}   bids[].usdcAmount   - USDC in 1e6 units.
+ * @param {bigint}   bids[].maxPrice     - Max price in USDC’s 1e6 format.
+ * @param {bigint}   bids[].priority     - Priority for tie-breaks.
+ * @param {bigint}   totalTokens         - Total tokens to sell in 1e18 units.
+ *
+ * @returns {{
+ *   finalPrice: bigint,  // Price per token in 1e6 USDC units (6 decimals)
+ *   allocations: {
+ *     [address: string]: {
+ *       tokens: bigint,      // Tokens allocated (18 decimals)
+ *       usedUSDC: bigint,    // USDC used (6 decimals)
+ *       refundedUSDC: bigint // USDC refunded (6 decimals)
+ *     }
+ *   }
+ * }}
+ */
+function runAuction(bids, totalTokens) {
+  // Constants for scaling
+  const TOKEN_SCALE = 1_000_000_000_000_000_000n; // 1 TOKEN = 1e18
+
+  // 1) Sort bids by maxPrice (desc), then priority (desc)
+  bids.sort((a, b) => {
+    if (b.maxPrice === a.maxPrice) {
+      return Number(b.priority - a.priority);
+    }
+    return Number(b.maxPrice - a.maxPrice);
+  });
+
+  // 2) Determine the clearing price
+  let tokensAccumulated = 0n;
+  let finalPrice = 0n;
+
+  for (const bid of bids) {
+    // tokensDemanded = (usdcAmount * TOKEN_SCALE) / maxPrice
+    const tokensDemanded = (bid.usdcAmount * TOKEN_SCALE) / bid.maxPrice;
+
+    tokensAccumulated += tokensDemanded;
+    if (tokensAccumulated >= totalTokens) {
+      finalPrice = bid.maxPrice;
```

Comment:
this is what I think is incorrect

Reply:
this is like the worst case. Because if I bid $300 for $30, you are only accumulating 10 tokens but in reality maybe you should have accumulated 30

Reply:
so you would run out of tokens earlier

Reply:
That's why I think the loop needs to be over each bid price first, and then over bids. It needs to be a double loop

Reply:
Yeah. I agree. thanks.